src/cargo/ops/cargo_rustc/links.rs

```diff
@@ -15,11 +15,23 @@ pub fn validate(deps: &PackageSet) -> CargoResult<()> {
         };
         match map.get(&lib) {
             Some(previous) => {
-                return Err(human(format!("native library `{}` is being linked \
-                                          to by more than one package, and \
-                                          can only be linked to by one \
-                                          package\n\n  {}\n  {}",
-                                         lib, previous, dep.package_id())))
+                if previous.name() == dep.package_id().name() {
```

Comment:
This'll need to check both `name` and `source_id` to ensure it's the same package but just different versions